controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,448 @@
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	// Get the data from Ironic and update HFS, the call to provisioner may fail if settings can't
+	// be retrieved for the bios_interface. Just log it and don't retry
+	if err = r.updateHostFirmwareSettings(prov, info); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Could not update hostFirmwareSettings")
+	}
+
+	for _, e := range info.events {
+		r.publishEvent(req, e)
+	}
+
+	// only run again on changes
+	return ctrl.Result{}, nil
+}
+
+// Get the firmware settings from the provisioner and update hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) updateHostFirmwareSettings(prov provisioner.Provisioner, info *rInfo) (err error) {
+
+	info.log.Info("retrieving firmware settings and saving to resource")
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		return errors.Wrap(err, "could not get firmware settings from provisioner")
+	}
+
+	// get or create a firmwareSchema to hold schema
+	firmwareSchema, err := r.getOrCreateFirmwareSchema(info, schema)
+	if err != nil {
+		return errors.Wrap(err, "could not get/create firmware schema")
+	}
+
+	// Set hostFirmwareSetting to use this schema
+	info.hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+		Namespace: firmwareSchema.ObjectMeta.Namespace,
+		Name:      firmwareSchema.ObjectMeta.Name}
+
+	if err = r.updateSpecAndStatus(info, currentSettings, firmwareSchema); err != nil {
+		return errors.Wrap(err, "could not update hostFirmwareSettings")
+	}
+
+	return nil
+}
+
+// Update the HostFirmwareSettings resource using the settings and schema from provisioner
+func (r *HostFirmwareSettingsReconciler) updateSpecAndStatus(info *rInfo, settings metal3v1alpha1.SettingsMap, schema *metal3v1alpha1.FirmwareSchema) (err error) {
+
+	if info.hfs.Status.Settings == nil {
+		info.hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	}
+
+	// Update Spec and Status
+	specDirty := false
+	for k, v := range settings {
+		// Some vendors include encrypted password fields, don't add these
+		if strings.Contains(k, "Password") {
+			continue
+		}
+		info.hfs.Status.Settings[k] = v
+
+		// Don't include setting in Spec if it cannot be changed (ReadOnly is set)
+		// or if it unique to this particular host, in order to prevent the unique
+		// settings from being copied to other hosts
+		if schema != nil {
+			if schemaSetting, ok := schema.Spec.Schema[k]; ok {
+				if (schemaSetting.ReadOnly != nil && *schemaSetting.ReadOnly == true) ||
+					(schemaSetting.Unique != nil && *schemaSetting.Unique == true) {
+					continue
+				}
+			}
+		}
+		// Copy setting to Spec only if doesn't exist as setting may have been set by user
+		if _, ok := info.hfs.Spec.Settings[k]; !ok {
+			info.hfs.Spec.Settings[k] = intstr.FromString(v)
+			specDirty = true
+		}
+	}
+
+	if specDirty {
```

Comment:
This is probably a good place to add some logging since we modify a resource

Reply:
Added log